server/schedule/operator/step.go

```diff
@@ -195,13 +199,133 @@ func (ap AddPeer) GetCmd(region *core.RegionInfo, useConfChangeV2 bool) *pdpb.Re
 		// The newly added peer is pending.
 		return nil
 	}
-	return createResponse(addNode(ap.PeerID, ap.ToStore), useConfChangeV2)
+	return createResponse(addNode(ap.PeerID, ap.ToStore, ap.IsWitness), useConfChangeV2)
+}
+
+// BecomeWitness is an OpStep that makes a peer become a witness.
+type BecomeWitness struct {
+	StoreID, PeerID uint64
+}
+
+// ConfVerChanged returns the delta value for version increased by this step.
+func (bw BecomeWitness) ConfVerChanged(region *core.RegionInfo) uint64 {
+	peer := region.GetStorePeer(bw.StoreID)
+	return typeutil.BoolToUint64(peer.GetId() == bw.PeerID)
+}
+
+func (bw BecomeWitness) String() string {
+	return fmt.Sprintf("change peer %v on store %v to witness", bw.PeerID, bw.StoreID)
+}
+
+// IsFinish checks if current step is finished.
+func (bw BecomeWitness) IsFinish(region *core.RegionInfo) bool {
+	if peer := region.GetStorePeer(bw.StoreID); peer != nil {
+		if peer.GetId() != bw.PeerID {
+			log.Warn("obtain unexpected peer", zap.String("expect", bw.String()), zap.Uint64("obtain-learner", peer.GetId()))
+			return false
+		}
+		return peer.IsWitness
+	}
+	return false
+}
+
+// CheckInProgress checks if the step is in the progress of advancing.
+func (bw BecomeWitness) CheckInProgress(ci ClusterInformer, region *core.RegionInfo) error {
+	if err := validateStore(ci, bw.StoreID); err != nil {
+		return err
+	}
+	peer := region.GetStorePeer(bw.StoreID)
+	if peer == nil || peer.GetId() != bw.PeerID {
+		return errors.New("peer does not exist")
+	}
+	return nil
+}
+
+// Influence calculates the store difference that current step makes.
+func (bw BecomeWitness) Influence(opInfluence OpInfluence, region *core.RegionInfo) {
+	to := opInfluence.GetStoreInfluence(bw.StoreID)
+
+	regionSize := region.GetApproximateSize()
+	to.RegionSize -= regionSize
+	// TODO: Confirm no performance issues
```

Comment:
yes, no performance issue, just like removing a peer

Reply:
Got it, thanks!